MNT deprecate attributes in Partial Least Squares module

marenwestermann opened the pull request:
#### Reference Issues/PRs
Towards #14312 
Closes #18764

#### What does this implement/fix? Explain your changes. 
Deprecation of the attributes x_mean_, y_mean_, x_std_, y_std_ in module _pls.py

#### Any other comments?
ping @adrinjalali

marenwestermann committed:
deprecate attributes

adrinjalali commented:
we're already deprecating a few attributes in 0.24 in these classes, and this PR deprecates more (which I'm in favor of), but there are a bunch of other attributes which can be made private as well, like, all of these maybe:

```
    x_weights_ : ndarray of shape (n_features, n_components)
        The left singular vectors of the cross-covariance matrices of each
        iteration.

    y_weights_ : ndarray of shape (n_targets, n_components)
        The right singular vectors of the cross-covariance matrices of each
        iteration.

    x_loadings_ : ndarray of shape (n_features, n_components)
        The loadings of `X`.

    y_loadings_ : ndarray of shape (n_targets, n_components)
        The loadings of `Y`.

    x_scores_ : ndarray of shape (n_samples, n_components)
        The transformed training samples.

    y_scores_ : ndarray of shape (n_samples, n_components)
        The transformed training targets.

    x_rotations_ : ndarray of shape (n_features, n_components)
        The projection matrix used to transform `X`.

    y_rotations_ : ndarray of shape (n_features, n_components)
        The projection matrix used to transform `Y`.

```

should we deprecate them and include them in the release? Kinda looks cleaner if they're all deprecated in the same release, WDYT @ogrisel  @NicolasHug

NicolasHug commented:
apart from the scores, the rest of the attributes are relevant and shouldn't be deprecated

adrinjalali commented:
what do you mean by relevant? I mean, what's the difference between `x_mean_` and and `x_rotations_` for instance in your opinion?

NicolasHug commented:
they all have specific meanings: https://scikit-learn.org/dev/modules/cross_decomposition.html#cross-decomposition

NicolasHug reviewed (commented):
Thanks @marenwestermann , looks good. 

we'll also need a quick test making sure these are deprecated. You can parametrize with all estimators and with all attributes.

marenwestermann committed:
add test

NicolasHug commented on sklearn/cross_decomposition/tests/test_pls.py:
```diff
@@ -488,6 +488,17 @@ def test_norm_y_weights_deprecation(Est):
         est.norm_y_weights
 
 
+@pytest.mark.parametrize('Est', (PLSRegression, PLSCanonical, CCA, PLSSVD))
```
```suggestion
# TODO: Remove test in 0.26
@pytest.mark.parametrize('Est', (PLSRegression, PLSCanonical, CCA, PLSSVD))
```

NicolasHug reviewed (approved):
thanks @marenwestermann , LGTM if green!

marenwestermann committed:
add remove test note

marenwestermann committed:
remove classes from ignored sets

NicolasHug commented:
@adrinjalali to expand on my (probably too) short reply above re the attributes: I think they're more relevant than the means and stds because they can be used for some specific purposes and they can also be used for inspection. The `rotations_` matrices are used to project the input into the transformed space, so they're basically the equivalent of the `components_` attributes for PCA. The `weights` are also related to projections, and the `loadings` are related to the regression (these estimators can do both `transform` and `predict`.)

adrinjalali commented:
> @adrinjalali to expand on my (probably too short) reply above re the attributes: I think they're more relevant than the means and stds because they can be used for some specific purposes and they can also be used for inspection. The `rotations_` matrices are used to project the input into the transformed space, so they're basically the equivalent of the `components_` attributes for PCA. The `weights` are also related to projections, and the `loadings` are related to the regression (these estimators can do both `transform` and `predict`.)

fair point, that's convincing :)

marenwestermann commented:
@NicolasHug I added the TODO note and removed the classes from the IGNORED set in ` sklearn/tests/test_docstring_parameters.py` as suggested by @glemaitre in #18764 (I hope it's correct). Do I also need to leave a note in ` doc/whats_new/v0.24.rst` saying that the attributes here are deprecated?

glemaitre commented on sklearn/cross_decomposition/tests/test_pls.py:
```diff
@@ -488,6 +488,18 @@ def test_norm_y_weights_deprecation(Est):
 
 
+# TODO: Remove test in 0.26
+@pytest.mark.parametrize('Est', (PLSRegression, PLSCanonical, CCA, PLSSVD))
```
```suggestion
@pytest.mark.parametrize('Estimator', (PLSRegression, PLSCanonical, CCA, PLSSVD))
```

glemaitre commented on sklearn/cross_decomposition/tests/test_pls.py:
```diff
@@ -488,6 +488,18 @@ def test_norm_y_weights_deprecation(Est):
 
+# TODO: Remove test in 0.26
+@pytest.mark.parametrize('Est', (PLSRegression, PLSCanonical, CCA, PLSSVD))
+@pytest.mark.parametrize('attr', ("x_mean_", "y_mean_", "x_std_", "y_std_"))
```
```suggestion
@pytest.mark.parametrize('attribute', ("x_mean_", "y_mean_", "x_std_", "y_std_"))
```

glemaitre commented on sklearn/cross_decomposition/tests/test_pls.py:
```diff
@@ -488,6 +488,18 @@ def test_norm_y_weights_deprecation(Est):
+    Y = rng.randn(10, 3)
+    est = Est().fit(X, Y)
+    with pytest.warns(FutureWarning, match=f"{attr} was deprecated"):
+        getattr(est, attr)
```
```suggestion
        getattr(estimato, attribute)
```

glemaitre commented on sklearn/cross_decomposition/tests/test_pls.py:
```diff
@@ -488,6 +488,18 @@ def test_norm_y_weights_deprecation(Est):
+    X = rng.randn(10, 5)
+    Y = rng.randn(10, 3)
+    est = Est().fit(X, Y)
+    with pytest.warns(FutureWarning, match=f"{attr} was deprecated"):
```
```suggestion
    with pytest.warns(FutureWarning, match=f"{attribute} was deprecated"):
```

glemaitre commented on sklearn/cross_decomposition/tests/test_pls.py:
```diff
@@ -488,6 +488,18 @@ def test_norm_y_weights_deprecation(Est):
+    rng = np.random.RandomState(0)
+    X = rng.randn(10, 5)
+    Y = rng.randn(10, 3)
+    est = Est().fit(X, Y)
```
```suggestion
    estimator = Estimator().fit(X, Y)
```

glemaitre reviewed (commented):
We will need an entry in the what's new to announce the deprecation of the parameters since they could be used.

Otherwise, LGTM

marenwestermann committed:
replace abbreviations with full words for improved readability

marenwestermann committed:
make entry in whats_new

marenwestermann commented:
I implemented the suggested changes and made an entry in the what's new section.

adrinjalali reviewed (approved):
Thanks @marenwestermann

adrinjalali commented:
ping @ogrisel for the release
